insecure/corruptible/factory.go

```diff
@@ -0,0 +1,239 @@
+
+import (
+	"context"
+	"fmt"
+	"io"
+
+	"github.com/golang/protobuf/ptypes/empty"
+	"github.com/rs/zerolog"
+	"google.golang.org/grpc"
+
+	"github.com/onflow/flow-go/insecure"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module/component"
+	"github.com/onflow/flow-go/network"
+)
+
+// ConduitFactory implements a corruptible conduit factory, that creates corruptible conduits and acts as their master.
+// A remote attacker can register itself to this conduit factory.
+// Whenever any corruptible conduit generated by this factory receives an event from its engine, it relays the event to this
+// factory, which in turn is relayed to the register attacker.
+// The attacker can asynchronously dictate the conduit factory to send messages on behalf of the node this factory resides on.
+type ConduitFactory struct {
+	*component.ComponentManager
+	logger                zerolog.Logger
+	codec                 network.Codec
+	myId                  flow.Identifier
+	adapter               network.Adapter
+	attackerObserveClient insecure.Attacker_ObserveClient
+}
+
+func NewCorruptibleConduitFactory(logger zerolog.Logger, chainId flow.ChainID, myId flow.Identifier, codec network.Codec) *ConduitFactory {
+	if chainId != flow.BftTestnet {
+		panic("illegal chain id for using corruptible conduit factory")
+	}
+
+	return &ConduitFactory{
+		ComponentManager: component.NewComponentManagerBuilder().Build(),
+		myId:             myId,
+		codec:            codec,
+		logger:           logger.With().Str("module", "corruptible-conduit-factory").Logger(),
+	}
+}
+
+// RegisterAdapter sets the Adapter component of the factory.
+// The Adapter is a wrapper around the Network layer that only exposes the set of methods
+// that are needed by a conduit.
+func (c *ConduitFactory) RegisterAdapter(adapter network.Adapter) error {
+	if c.adapter != nil {
+		return fmt.Errorf("could not register a new network adapter, one already exists")
+	}
+
+	c.adapter = adapter
+
+	return nil
+}
+
+// NewConduit creates a conduit on the specified channel.
+// Prior to creating any conduit, the factory requires an Adapter to be registered with it.
+func (c *ConduitFactory) NewConduit(ctx context.Context, channel network.Channel) (network.Conduit, error) {
+	if c.adapter == nil {
+		return nil, fmt.Errorf("could not create a new conduit, missing a registered network adapter")
+	}
+
+	child, cancel := context.WithCancel(ctx)
+
+	con := &Conduit{
+		ctx:     child,
+		cancel:  cancel,
+		channel: channel,
+		master:  c,
+	}
+
+	return con, nil
+}
+
+func (c *ConduitFactory) ProcessAttackerMessage(stream insecure.CorruptibleConduitFactory_ProcessAttackerMessageServer) error {
+	for {
+		select {
+		case <-c.ComponentManager.ShutdownSignal():
+			if c.attackerObserveClient != nil {
+				_, err := c.attackerObserveClient.CloseAndRecv()
+				if err != nil {
+					c.logger.Fatal().Err(err).Msg("could not close processing stream from attacker")
+					return err
+				}
+			}
```

Comment:
```suggestion
			}
			returtn nil
```

Reply:
Thanks for catching this, applied in https://github.com/onflow/flow-go/pull/2033/commits/e55c51e2c7529904ab0abdd7ade71b60577d96a0.